src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -947,27 +1042,44 @@ private bool TryProcessEventCore(CompilationEvent e, AnalysisScope analysisScope
         /// False, otherwise.
         /// </returns>
-        private bool TryProcessSymbolDeclared(SymbolDeclaredCompilationEvent symbolEvent, AnalysisScope analysisScope, AnalysisState analysisStateOpt, CancellationToken cancellationToken)
+        private async Task<EventProcessedState> TryProcessSymbolDeclaredAsync(SymbolDeclaredCompilationEvent symbolEvent, AnalysisScope analysisScope, AnalysisState analysisStateOpt, CancellationToken cancellationToken)
         {
             try
             {
                 // Attempt to execute all analyzer actions.
-                var success = true;
+                var processedState = EventProcessedState.Processed;
                 var symbol = symbolEvent.Symbol;
                 var isGeneratedCodeSymbol = IsGeneratedCodeSymbol(symbol);
-                if (!AnalysisScope.ShouldSkipSymbolAnalysis(symbolEvent) &&
+
+                var skipSymbolAnalysis = AnalysisScope.ShouldSkipSymbolAnalysis(symbolEvent);
+                var skipDeclarationAnalysis = AnalysisScope.ShouldSkipDeclarationAnalysis(symbol);
+                var hasPerSymbolActions = AnalyzerActions.SymbolStartActionsCount > 0 && (!skipSymbolAnalysis || !skipDeclarationAnalysis);
+
+                AnalyzerActions perSymbolActions = hasPerSymbolActions ?
+                    await GetPerSymbolAnalyzerActionsAsync(symbol, analysisScope, analysisStateOpt, cancellationToken).ConfigureAwait(false) :
+                    null;
+
+                if (!skipSymbolAnalysis &&
                     !TryExecuteSymbolActions(symbolEvent, analysisScope, analysisStateOpt, isGeneratedCodeSymbol, cancellationToken))
                 {
-                    success = false;
+                    processedState = EventProcessedState.NotProcessed;
                 }
 
-                if (!AnalysisScope.ShouldSkipDeclarationAnalysis(symbol) &&
-                    !TryExecuteDeclaringReferenceActions(symbolEvent, analysisScope, analysisStateOpt, isGeneratedCodeSymbol, cancellationToken))
+                if (!skipDeclarationAnalysis &&
+                    !TryExecuteDeclaringReferenceActions(symbolEvent, analysisScope, analysisStateOpt, isGeneratedCodeSymbol, perSymbolActions, cancellationToken))
                 {
-                    success = false;
+                    processedState = EventProcessedState.NotProcessed;
                 }
 
-                return success;
+                ImmutableArray<DiagnosticAnalyzer> subsetProcessedAnalyzers = default;
+                if (processedState.Kind == EventProcessedStateKind.Processed &&
+                    hasPerSymbolActions &&
+                    !TryExecuteSymbolEndActions(perSymbolActions, symbolEvent, analysisScope, analysisStateOpt, cancellationToken, out subsetProcessedAnalyzers))
+                {
+                    processedState = subsetProcessedAnalyzers.IsDefaultOrEmpty ? EventProcessedState.NotProcessed : EventProcessedState.CreatePartiallyProcessed(subsetProcessedAnalyzers);
```

Comment:
>EventProcessedState.CreatePartiallyProcessed(subsetProcessedAnalyzers) [](start = 116, length = 70)

Can you please explain how and why we track this (partial) set of processed analyzers? It seems to be key to this change but I don't know why.

Reply:
We do this primarily to improve responsiveness for IDE cases where it executes analyzers one at a time on each open source file at a time. We have to track the symbol completion state per-analyzer or we need to wait till all analyzers have completed analysis of a given symbol and it's members before we can execute the symbol end actions.